[#35] 로그인한 유저 인증 기능 추가

Srltas opened the pull request:
## 📌 Summary
Access Token을 통해 사용자 인증 후, 인증된 사용자 정보를 HttpSession에 저장하여 Controller에서 사용할 수 있도록 하는 로직을 추가했습니다. 

## 📝 Description
AuthenticationFilter
 - Authorization 헤더에서 Bearer 토큰을 추출하여, 이를 SessionStorage에서 조회하여 사용자 정보를 가져옵니다.
 - 조회된 사용자 정보(UserSessionDTO)를 HttpSession에 저장하여 이후 요청에서 사용할 수 있도록 설정합니다.
 - 헤더에 Authorization 또는 저장되어 있는 세션이 조회되지 않으면, 더 이상 요청 작업을 진행하지 않고 에러를 내면서 종료합니다.

SessionStorage
 - 인터페이스로 분리하여 구현한 이유는 추후 Redis와 같은 외부 저장소를 도입할 때 기존 코드를 최소한으로 변경하면서도 손쉽게 확장할 수 있도록 했습니다.
 - 현재는 구체적인 구현 없이 인터페이스만 정의해 두었고, 이후 Redis를 학습한 뒤 이를 기반으로 구현할 계획입니다.

Controller
 - HttpSession을 통해 사용자 정보를 받아 사용할 수 있도록 수정했습니다.

## ✅ Checklist
- [x] 코딩 스타일 가이드를 준수했습니다.
- [x] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [x] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

Srltas committed:
feat: 로그인한 유저 인증 기능 추가

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/NeighborhoodVerificationController.java:
```diff
@@ -34,12 +35,16 @@ public class NeighborhoodVerificationController {
-		@RequestBody @Valid NeighborhoodVerificationRequest neighborhoodVerificationRequest,
-		@Parameter(hidden = true) @SessionAttribute(name = "login_user_id", required = false) Long userId) {
+		@RequestBody @Valid NeighborhoodVerificationRequest neighborhoodVerificationRequest, HttpSession session) {
+		UserSessionDTO userSession = (UserSessionDTO)session.getAttribute("USER_SESSION");
```
"USER_SESSION" 은 constant 로 선언하는게 좋아 보이네요.

Srltas replied:
코드의 가독성을 높이고, 중복된 문자열로 인해 발생할 수 있는 실수를 방지하기 위해서 상수로 설정하는 것이 좋을거 같습니다!
더 나아가 "USER_SESSION" 문자를 Filter에서만 사용하는 것이 아니라 Controller에서도 사용하고 있어 어느 계층에도 종속되지 않고 독립적으로 관리 될 수 있도록 common 패키지에 상수 클래스를 만들어 수정했습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,49 @@
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class AuthenticationFilter implements Filter {
```
OncePerRequestFilter 에 대해서 알아 보시겠어요???

Srltas replied:
`OncePerRequestFilter`는 스프링에서 제공해주는 Filter입니다.
같은 요청이 FORWARD나 INCLUDE를 통해 처리되는 경우, filter가 한 요청에 대해 여러 번 실행될 수 있습니다. 하지만, `OncePerRequestFilter`는 동일한 요청에 대해서는 한 번만 실행되도록 보장합니다.

해당 필터는 로그인한 유저의 요청인지 확인하는 필터로 한 번만 실행 되어도 문제가 없어 `OncePerRequestFilter`를 사용하는 것이 좋을거 같습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,49 @@
+		String authorizationHeader = request.getHeader("Authorization");
+		String token = extractToken(authorizationHeader);
+
+		if (token != null) {
```
null 일 경우 에러 처리를 안하는건 controller 에서 null 이면 에러를 내릴 거라서 그런건가요?
그럼 모든 API 가 생길 때마다 컨트롤러에서는 해당 로직이 있어야 겠네요???

Srltas replied:
잘못된 설계인거 같습니다😥
Filter에서 null 에러 처리를 하여 Controller까지 요청을 보내지 않고 끝내는 것이 더 적절할거 같습니다.
Filter에서 처리할 수 있도록 수정하겠습니다.

Srltas committed:
refactor: USER_SESSION 상수 클래스로 만들어 사용

Srltas committed:
refactor: 인증 관련 상수 클래스 생성

- Header에 인증과 키워드들을 상수 클래스로 정의하여 사용

Srltas committed:
refactor: OncePerRequestFilter를 상속하도록 수정

 - AuthenticationFilter는 해당 요청이 로그인된 사용자의 요청인지 확인하는 필터입니다
 - 해당 필터는 요청 당 한 번만 실행되면 되므로 한 번만 실행될 수 있도록 AuthenticationFileter를 상속했습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/SessionAttribute.java:
```diff
@@ -0,0 +1,10 @@
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = PRIVATE)
```
lombok 의 UtilityClass 를 사용해도 될것 같아요

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/AuthConstants.java:
```diff
@@ -0,0 +1,12 @@
+package com.srltas.runtogether.common;
```
해당 값들은 web 에서만 사용하는거 아닌가요?
굳이 common 에서 관리 할 필요가 있을까요???

Srltas replied:
이 상수들이 여러 계층에서 사용될 수 있다고 생각하여 특정 계층에 종속되지 않도록 공통 패키지에 둬서 재사용하고자 했습니다.

하지만, 현재 상수들이 실제로 사용되는 곳은 웹 계층에만 한정되어 있고, 다른 계층에서도 이 상수들이 필요할 가능성이 작을 것이라는 생각도 듭니다. 

우선은 웹 계층에만 연관이 되어 있기 때문에 웹 패키지로 이동하여 구조적 일관성과 응집성을 높이는 방향으로 변경하도록 하겠습니다.

Srltas committed:
refactor: filter에서 token이 null 처리하도록 수정

- filter에서 token이 null인 경우 더 이상 요청을 진행하지 않고 에러를 반환합니다.
- controller에서는 filter에서 인증 처리를 했기 때문에, 더 이상 null 체크는 하지 않습니다.

Srltas committed:
refactor: lombok의 @UtilityClass 사용

Srltas committed:
refactor: 세션, 헤더 권한 유틸클래스 패키지 변경

Srltas committed:
test: filter, controller 테스트 코드 생성 및 수정

f-lab-bradley commented on src/main/java/com/srltas/runtogether/config/WebConfig.java:
```diff
@@ -0,0 +1,26 @@
+	public FilterRegistrationBean<AuthenticationFilter> sessionFilterRegistration() {
+		FilterRegistrationBean<AuthenticationFilter> registrationBean = new FilterRegistrationBean<>();
+		registrationBean.setFilter(new AuthenticationFilter(sessionStorage));
+		registrationBean.addUrlPatterns("/neighborhood/verification");
```
이 부분이 Controller 와 분리 되어 있네요.
만약 API 의 URL 이 바뀐다면 항상 이 부분을 고쳐 줘야 하는데, 유지보수성 측면에서 힘들지 않을까요?

Srltas replied:
코드에 직접 URL을 넣는 것 보단 프로퍼티 파일과 같은 외부 설정에서 URL을 관리하면, 코드 수정 없이 설정만 변경할 수 있어 URL 경로가 바뀔 때도 소스 코드를 수정할 필요가 없고, 설정 파일을 통해 유연하게 대처할 수 있습니다.

혹시 이 외에 더 나은 방법이 있을까요?

f-lab-bradley replied:
그냥 한 곳에 모아서 Constant 처럼 관리하는게 낫지 않을까요?
지금 구조라면 url 이 바뀌면 controller 도 바뀌어야 하고 Filter configuration 도 찾아서 바꿔야 할 것 같은데 관리가 힘들 것 같아서요.

Srltas replied:
URL을 관리하는 유틸 클래스를 하나 만들어 사용하도록 수정했습니다.

f-lab-bradley commented on src/test/java/com/srltas/runtogether/adapter/in/NeighborhoodVerificationControllerTest.java:
```diff
@@ -20,46 +20,53 @@
-	@DisplayName("verifyNeighborhood 메서드에 대한 Parameterized 테스트")
-	void verifyNeighborhood_ShouldReturnOkResponse(NeighborhoodVerificationRequest request, Long userId) {
+	@MethodSource("provideRequestsForSuccess")
+	@DisplayName("세션에 사용자 정보가 있을 때 동네 인증 성공 여부 확인")
```
굳 👍

Srltas committed:
refactor: URL constant로 관리

- Controller와 Filter 두 곳에 URL 관리의 편의성을 위해 constant 클래스를 만들어 한 곳에서 관리합니다

Srltas committed:
feat: 임시로 사용하는 SessionStorage 구현체

- 로그인 기능을 개발하기 전까지 테스트용으로 사용할 SessionStorage 구현체입니다

Srltas commented:
로그인 기능을 개발하기 전까지 테스트용으로 사용할 임시 SessionStorage 구현체를 만들어 적용했습니다.

f-lab-bradley reviewed (approved):
수고하셨습니다~

Srltas committed:
refactor: Filter 코드 리펙토링

- 저장소에서 유저가 존재하는지 확인하는 메소드(authenticateUser)가 확인 뿐만 아니라 request에 유저 정보를 넣는 역할까지 맞고 있어 책임 분리
- 역할을 명확하게 하기 위해 기존 메소드의 이름을 authenticateUser -> isNullUserSessionDTO로 변경
- doFilterInternal에서 request에 유저 정보를 넣어 Controller에게 전달하는 역할 수행
- 토큰이 null인 경우 저장소에서 유저를 확인하지 않고, 바로 예외 처리를 할 수 있도록 로직 수정

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/filter/AuthenticationFilter.java:
```diff
@@ -29,24 +29,21 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
-		HttpSession session = req.getSession(true);
-		session.setAttribute(USER_SESSION, userSessionDTO);
-		return true;
+	private boolean isNullUserSessionDTO(UserSessionDTO userSessionDTO) {
```
흠.. 흠.... 뭔가 뭔가인데.. 음......
일단....알겠습니다...ㄷㄷ;;;

Srltas replied:
아.. 음... 먼가 어색한가요?..
메소드 이름 때문일까요?.. 아니면 다른 이유가 있나요?😢

f-lab-bradley replied:
뭔가 크게 개선 된 느낌은 없는 것 같아요.
나중에 코드 다듬으면서 다시 봐도 될 것 같네요

Srltas replied:
넵!.. 로그인 기능을 넣을 때 다시 한번 코드를 다듬어 보겠습니다.
